### DISCUSSION TITLE
Example is not working

### DISCUSSION THREAD

**Discussion by janwirth** [pointfreeco/swift-composable-architecture#3535] [General]
### Description

I copy pasted the reducer & view examples from the documentation and I get the following errors (among others):

<img width="665" alt="image" src="https://github.com/user-attachments/assets/8d070a46-cdd6-47a6-920a-750520cc6eee" />

<img width="278" alt="image" src="https://github.com/user-attachments/assets/692d440c-d0e7-4c7a-8100-8e0cf7754548" />

I also had to enable some package macros which was not part of the documentation / installation instruction.

The library seems to have a lot of dependencies too.

I'm not super experienced with swift, I have ~10 years of web dev experience.
Generally I try to avoid installing dependencies that seem too heavy.

I'm leaving this here also as a first-time user experience, and I will try to understand the low-level domain (Observables etc.) better, before opting into a rather large library.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

It compiles.

### Actual behavior

See compiler error screenshots.

### The Composable Architecture version information

Pulled from main today.

### Destination operating system

iOS / MacOs

### Xcode version information

<img width="196" alt="image" src="https://github.com/user-attachments/assets/b7a2a052-de42-4eab-a625-d7786463411f" />

**Comment by ysmaliak** [comment]
Hi @janwirth! It looks like you have empty `State` and `Action` in your reducer. You can check the [Examples](https://github.com/pointfreeco/swift-composable-architecture/tree/main/Examples) folder; it contains finished apps that should work just fine.

However, I would recommend checking out version `1.16.1` instead of the current `1.17.0` because of some issues with shared properties that I have encountered.

**Comment by ysmaliak** [comment]
I would also recommend checking out videos at https://www.pointfree.co; they have a lot of good episodes about almost everything you would need to know.

**Comment by mbrandonw** [comment]
I am going to convert this to a discussion since it's not an issue with the library. You can provide a project that demonstrates the problem you are seeing if you want others to help.

**Comment by janwirth** [comment]
Sorry, I haven't responded to this yet.

I overlooked that the other code was commented out - this was part of a hackathon on a pet project that I'm raising to v2 and I was sick, so tired / not super focussed.

That being said, **I think having a self-contained copy-and-paste example on top of documentation is always a good idea.** This helps communicate the way the library works and makes it easy to get started.

Beyond this, why is this library installing so many dependencies and enabling various macros?

The flux architecture is rather simple to implement in my experience, but **when I checkout SCA in the Swift PM it appears to be fetching dozens of dependencies.**

**Reply by stephencelis** [reply]
> That being said, **I think having a self-contained copy-and-paste example on top of documentation is always a good idea.** This helps communicate the way the library works and makes it easy to get started.

The documentation you're looking at is incremental. If you scroll up you will see the full versions of `State` and `Action`, and if you implement things incrementally you should have something compiling.

We use Apple's documentation convention of `/* ... */` and `// ...` comments to denote code that has been collapsed, as repeating it a bunch of places would bloat the README and introduction. In Swift and in _general_ I think it's quite common for documentation examples to rarely fully compile in isolation.

> Beyond this, why is this library installing so many dependencies and enabling various macros?

Most of these dependencies come from the pointfreeco org and were originally incubated for TCA functionality but extracted for more general reuse. Many have become popular outside the original TCA use case. Feel free to evaluate each dependency to determine its usefulness.

> The flux architecture is rather simple to implement in my experience, but **when I checkout SCA in the Swift PM it appears to be fetching dozens of dependencies.**

TCA offers a lot of functionality that flux does not. I'd recommend evaluating the library's offerings and then decide if it is something you want to use. You are always free to implement flux manually in your project if you find you don't need TCA's functionality.

**Reply by janwirth** [reply]
I did understand that it is incremental, I'm trying to communicate that a full, self-contained example is the default pattern that I expect, coming from the JS ecosystem.

Thank you for responding to my concerns. I still have a lot to learn in the Swift Ecosystem, and I have PTSD from bloat and framework churn in the JS ecosystem, so I generally prefer to have very light dependencies instead of heavy ones.

For me it is hard to evaluate how everything fits together, what the added value compared to flux is.

**Comment by janwirth** [comment]
That being said, I appreciate your swift responses, it speaks to the quality of your work, but the apparent bloat makes me skeptical. 

Thank you for the pointers, I checked pointfree before and I like the nod to tacit-style programming. I'm a big fan of elm and functional programming in general.

I will see what I can learn from the free point-free episodes.

So far I've been looking at the ObjectBinding / state management episodes and wondering about why this is chosen over the EnvironmentObject for global state.

I think a search feature / RAG use case could really enhance the accessibility of your content.

**Comment by janwirth** [comment]
Also it's hard for me to understand who this documentation is aimed at? Entry-level engineers? People that want to build their swift apps out to cross platform?

I personally never used an online course or a tutorial in my life, so I'm perhaps not the target group. I usually have very specific goals and look for very relevant information for the problem at hand, learning by doing.

**Reply by stephencelis** [reply]
> Also it's hard for me to understand who this documentation is aimed at? Entry-level engineers? People that want to build their swift apps out to cross platform?

Our documentation is aimed at Swift developers that are familiar with Xcode and the Apple ecosystem. We document what the library is solving for and how its APIs work. We avoid documenting things that are general Swift and Apple development knowledge and that can be encountered or Googled outside TCA usage, like the need to enable a package's macros.

**Comment by janwirth** [comment]
Thank you so much PF team, I already took too much of your time! I learnt a lot
